### PR TITLE
Add exhaustive calibration fallback

### PR DESCRIPTION
**Pull request by csherry04**
Added `cv2.CALIB_CB_EXHAUSTIVE` to the SB fallback path

Upon testing was finding that it was not significantly slower and in many cases was faster due to less upsampled retries. 

Compiled and tested lots of videos with the following results:
ACL are field videos where primary detector struggled, new fallback was better but still missing some, exhaustive was perfect

Comprehensive is my own set of calibrations with partial boards, distracting checkerboard-esque items to try and slow down exhaustive, etc

Utah set was passing with primary detector on my machine but not on others

| Dataset | Mode | Cases | Pass | Fail | Total Time |
|---|---:|---:|---:|---:|---:|
| ACL | Current | 18 | 13 | 5 | 22.294s |
| ACL | Exhaustive | 18 | 18 | 0 | 17.989s |
| Comprehensive | Current | 10 | 2 | 8 | 14.017s |
| Comprehensive | Exhaustive | 10 | 2 | 8 | 14.512s |
| LabValidation | Current | 90 | 90 | 0 | 31.144s |
| LabValidation | Exhaustive | 90 | 90 | 0 | 31.356s |
| Utah | Current | 4 | 4 | 0 | 1.606s |
| Utah | Exhaustive | 4 | 4 | 0 | 1.613s |
| **All** | **Current** | **122** | **109** | **13** | **69.061s** |
| **All** | **Exhaustive** | **122** | **114** | **8** | **65.470s** |

Worth noting everything that failed with exhaustive was intended to fail.

Even in cases with no board, partial board, and distracting items, the exhaustive change was on average less than .1 seconds slower per video. 

Included in PR is a new calibration testing file that runs through many of these trials and asserts correct behavior as well as low reprojection errors as well as a new more simple test in the main E2E testing file. Updated submodule to contain relevant videos

### CODE REVIEW

File: tests/test_calibration.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+from utilsChecker import (
+    calcExtrinsicsFromVideo,
+    generate3Dgrid,
+    loadCameraParameters,
+    rotateIntrinsics,
+)
+
+
+STANDARD_CHECKERBOARD_PARAMS = {
```
Comment (carmichaelong) [inline tests/test_calibration.py]: for maintainability by future devs, suggest names:
- `DEFAULT_CHECKERBOARD_PARAMS`
- `LABVALIDATION_CHECKERBOARD_PARAMS`

File: tests/test_calibration.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+sys.path.append(repoDir)
+
+from utilsChecker import (
+    calcExtrinsicsFromVideo,
+    generate3Dgrid,
+    loadCameraParameters,
+    rotateIntrinsics,
+)
+
+
```
Comment (carmichaelong) [inline tests/test_calibration.py:23]: some section headers could be helpful. some suggestions:
```
# ----  Checkerboard / fixture constants ----
# ---- Unpatched cv2 handles ----
# ---- Helpers ----
# ---- Tests ----
```

File: tests/test_calibration.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    'dimensions': (5, 4),
+    'squareSize': 35.0,
+}
+LABVALIDATION_CHECKERBOARD_PARAMS = {
+    'dimensions': (11, 8),
+    'squareSize': 60.0,
+}
+MAX_MEAN_REPROJECTION_ERROR_PX = 0.5
+
+ORIGINAL_FIND_CHESSBOARD_CORNERS = cv2.findChessboardCorners
```
Comment (carmichaelong) [inline tests/test_calibration.py]: Consider moving these just below the fixtures

File: tests/test_calibration.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+ORIGINAL_FIND_CHESSBOARD_CORNERS_SB_WITH_META = cv2.findChessboardCornersSBWithMeta
+ORIGINAL_CORNER_SUB_PIX = cv2.cornerSubPix
+
+CALIBRATION_FIXTURE_DIR = os.path.join(
+    thisDir,
+    'opencap-test-data',
+    'Data',
+    'calibration-fixtures',
+)
+IPHONE13_MODEL = 'iPhone13,3'
```
Comment (carmichaelong) [inline tests/test_calibration.py]: to be consistent with below, `IPHONE13_3_MODEL`

File: tests/test_calibration.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+        iphone_model,
+        INTRINSICS_FOLDER,
+        'cameraIntrinsics.pickle',
+    )
+    camera_params = loadCameraParameters(intrinsics_path)
+    return rotateIntrinsics(camera_params, str(video_path))
+
+
+def input_media_dir(tmp_path):
+    media_dir = (
```
Comment (carmichaelong) [inline tests/test_calibration.py]: consistency: i'd lean towards using `os.path.join()`

File: tests/test_calibration.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+    return flags
+
+
+def mean_reproj_error(camera_params, checkerboard_params, corners, image_shape):
+    if camera_params is None or corners is None or image_shape is None:
+        return None
+
+    image_width = image_shape[1]
+    camera_image_width = float(np.squeeze(camera_params['imageSize'])[1])
+    scale = image_width / camera_image_width
```
Comment (carmichaelong) [inline tests/test_calibration.py:188]: could you add a (short) comment here to explain why the scaling is necessary?

File: tests/test_calibration.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+    checkerboard_params,
+    iphone_model,
+    tmp_path,
+    monkeypatch,
+    detector_mode,
+):
+    staged_video_path = stage_video(tmp_path, video_path)
+    camera_params = load_intrinsics(staged_video_path, iphone_model)
+    calls = {'primary': 0, 'fallback': 0}
+    production_flags = []
```
Comment (carmichaelong) [inline tests/test_calibration.py]: what are the intent of `production_flags` and `detector_flags`? perhaps a different name for these could make the flow of this helper function self-explanatory

File: tests/test_calibration.py (hunk 3 above)
Comment (carmichaelong) [reply tests/test_calibration.py]: a rename for these variables could help devs understand intent. for instance:
`ORIGINAL_FIND_CHESSBOARD_CORNERS` -> `UNPATCHED_CV2_FIND_CHESSBOARD_CORNERS`

File: utilsChecker.py
Comment (carmichaelong) [inline utilsChecker.py]: this comment needs to be updated now that `EXHAUSTIVE` is included

File: tests/test_calibration.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+        monkeypatch,
+        detector_mode='primary_only',
+    )
+
+    assert_extrinsics(result)
+    assert mean_error < MAX_MEAN_REPROJECTION_ERROR_PX, (
+        fixture_name,
+        mean_error,
+    )
+    assert calls['primary'] == 1
```
Comment (carmichaelong) [inline tests/test_calibration.py]: would a case that needs an upsampled image pass this test? (https://github.com/opencap-org/opencap-core/blob/e1a44f3beab74e42832bec0d2aa50087df655908/utilsChecker.py#L457)

i think OK to keep this as a test but add one in-line comment about this being this specifically checking for 1

File: tests/test_main.py [hunk 9]
```diff
@@ -10,6 +12,37 @@
 sys.path.append(repoDir)
 from main import main
 
+CALIBRATION_FIXTURE_DIR = os.path.join(
```
Comment (carmichaelong) [inline tests/test_main.py]: i think some of these fixtures are the same as in `test_calibration`. could be helpful to add a `conftest.py` file at this point for shared fixtures

**Review by carmichaelong — CHANGES_REQUESTED** [review]
Thanks @csherry04. Overall tests are looking good. I have a bunch of comments to help other devs for maintainability.

One thing not in here are unit tests for some of the new functions (and tests for checking the correctness of the extrinsics, in addition to if the board was detected). I'd suggest leaving that out of this PR, but we should make an issue to track that and add those soon.

File: tests/test_calibration.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+    calls = {'primary': 0, 'fallback': 0}
+    production_flags = []
+    detector_flags = []
+    detected = {'corners': None, 'image_shape': None}
+
+    def primary_detector(*args, **kwargs):
+        calls['primary'] += 1
+        found, corners = ORIGINAL_FIND_CHESSBOARD_CORNERS(*args, **kwargs)
+        if found:
+            detected['corners'] = corners.copy()
```
Comment (carmichaelong) [inline tests/test_calibration.py]: might be worth at least a comment (unless there's a quick way to defend against this) that the patch grabs the `findChessboardCorners()` output, but that other steps could manipulate the results further in `calcExtrinsics()`: https://github.com/opencap-org/opencap-core/blob/e1a44f3beab74e42832bec0d2aa50087df655908/utilsChecker.py#L504

or if indeed is being accounted for in the test, would be helpful to document why it works here

**Review by carmichaelong — CHANGES_REQUESTED** [review]

File: tests/test_calibration.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+IPHONE17_3_MODEL = 'iPhone17,3'
+INTRINSICS_FOLDER = 'Deployed'
+
+ACL_EXHAUSTIVE_FALLBACK_VIDEO = os.path.join(
+    CALIBRATION_FIXTURE_DIR,
+    'acl',
+    'exhaustive_fallback_success',
+    'acl_exhaustive_only_success.qt',
+)
+UTAH_PRODUCTION_VIDEO = os.path.join(
```
Comment (carmichaelong) [inline tests/test_calibration.py]: i don't think "production" is the right description here for the test. maybe `UTAH_VIDEO` or start with `UTAH_FIXTURES` in case we add more from that dataset

File: tests/test_calibration.py [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+        mean_error,
+    )
+    assert calls['primary'] == 1
+    assert calls['fallback'] == 0
+    assert production_flags == []
+    assert detector_flags == []
+
+
+# Utah vid should succeed through the production fallback route (and in many cases via primary detector)
+def test_utah_production_calibrates(tmp_path, monkeypatch):
```
Comment (carmichaelong) [inline tests/test_calibration.py]: double check function name here. note maybe we should move away from "production", but also `*_calibrates()` isn't parallel with other naming

**Review by carmichaelong — CHANGES_REQUESTED** [review]
also double checking: i don't think these currently include some of the tests that exercise a rotated board. but this could be lumped into the issue that we should open above and a later PR too

File: tests/test_calibration.py (hunk 2 above)
Comment (csherry04) [reply tests/test_calibration.py:23]: added section headers

File: tests/test_calibration.py (hunk 6 above)
Comment (csherry04) [reply tests/test_calibration.py:188]: fixed

File: tests/test_calibration.py (hunk 3 above)
Comment (csherry04) [reply tests/test_calibration.py]: resolved

File: tests/test_calibration.py (hunk 4 above)
Comment (csherry04) [reply tests/test_calibration.py]: decided might be easiest to just write these inline, let me know what you think

File: tests/test_calibration.py (hunk 5 above)
Comment (csherry04) [reply tests/test_calibration.py]: fixed

File: tests/test_calibration.py (hunk 7 above)
Comment (csherry04) [reply tests/test_calibration.py]: I have reworked this. Originally was meant to store flags that were requested by the production code vs the ones that were actually run in these tests. However this is almost always the same thing aside from the case where it simulates old fallback behavior. Structure is now consolidated to just store fallback_flags and there is a special case for that one instance

File: utilsChecker.py
Comment (csherry04) [reply utilsChecker.py]: resolved

File: tests/test_calibration.py (hunk 8 above)
Comment (csherry04) [reply tests/test_calibration.py]: This was because the test did in fact pass with no retries on my machine but bc thats not really the goal of the test I made this more loose to just ensure it passes within the initial 4 attempts and doesn't rely on fallback

File: tests/test_main.py (hunk 9 above)
Comment (csherry04) [reply tests/test_main.py]: have started this, will update it more in the pr that has more fixtures in main test file

File: tests/test_calibration.py (hunk 10 above)
Comment (csherry04) [reply tests/test_calibration.py]: these corner values were updated later by the sub pixel refinement that was also monkeypatched. However I made this more clear by storing the pre and post refined values separately. As far as I can tell that is the only modification that is made later?

File: tests/test_calibration.py (hunk 11 above)
Comment (csherry04) [reply tests/test_calibration.py]: resolved

File: tests/test_calibration.py (hunk 12 above)
Comment (csherry04) [reply tests/test_calibration.py]: resolved

**Review by carmichaelong — APPROVED** [review]
